[Feature] Create bundled and non-bundled (downloadable dependencies/packages) versions

BobbyESP opened the pull request:
## Pull Request: Move to Downloadable Plugins

### Description
This pull request addresses the following changes:

- **Transition from Pre-packaged Libraries:** We are moving away from using pre-packaged libraries in favor of downloadable plugins. This change aims to enhance flexibility, maintainability, and customization options for our project.

### Changes Made
- All of #254 
- Moved to Gradle 8.1.4
- Moved to Kotlin 1.9.20
- Created a Jetpack Compose app (that can be removed later) for testing. I feel more comfortable with Compose than XML.
- Migrated to Kotlin DSL (Gradle config)

### Benefits
- **Flexibility:** Downloadable plugins allow users to choose and update specific components independently (even if Python and FFmpeg are mandatory).
- **Reduced Package Size:** This transition helps in minimizing the overall package size (and also apps size that depend on the lib), resulting in faster downloads and reduced resource consumption.
- **Easier Updates:** Users can easily update individual plugins without waiting for a complete library update.

### Testing
- Made just manual tests.

### Additional Notes
- None by now

### Related Issues
- #248 

Please review and provide feedback on the proposed changes.

BobbyESP committed:
Added Kotlinx-serialization (WIP)

BobbyESP committed:
Updated to Kotlin 1.8.20 and deleted unnecessary code in ZipUtils.
Moved plugins declaration

BobbyESP committed:
Fully moved from Jackson to Kotlinx Serialization library

BobbyESP committed:
Bumped up version of the lib and updated libraries

BobbyESP committed:
Kind of merged the last commit of the repo; added MASTER downloadable yt-dlp

BobbyESP committed:
Merge branch 'master' into master_kotlinx_serialization

BobbyESP committed:
refactor: Moved files to their respective folders to keep smth like a clean architecture

BobbyESP committed:
feat: Added file downloader with progress
feat(NEW APP): Created a Jetpack Compose test app

BobbyESP committed:
chore: Prepared libraries for Gradle update

BobbyESP committed:
chore: Updated Gradle to 8.1.4
perf: Improved StreamProcessExtractor

BobbyESP committed:
chore: Updated Kotlin to 1.9.20

BobbyESP committed:
chore: Moved from deprecated `packagingOptions` to `packaging`

BobbyESP committed:
refactor: Moved variable names (some of them) to an `object`
feat: Started plugin downloader feature

BobbyESP committed:
fix: Fixed yt-dlp not working because of Kotlin update

BobbyESP committed:
refactor: Added comments to the main functions of the library

BobbyESP reviewed (commented):
Not yt-dlp but updates of it in commit [a71509](https://github.com/yausername/youtubedl-android/pull/256/commits/a71509cb1ac099241d6ef8ed4a5ecd11373dbd5e)

BobbyESP committed:
feat: Finished the PluginsDownloaderImpl
refactor: Optimized imports in `library` module

BobbyESP committed:
feat(WIP) ⚠️NOT TESTED YET!: Plugin download on library init

BobbyESP committed:
feat: Finished plugins support (needs testing)

BobbyESP committed:
feat: Added callback support to the `init` function

BobbyESP committed:
Added `null` to init function in Hava (don't know how callbacks work in Java sorry)

BobbyESP commented:
The plugins support is almost finished; I have to change some things related to the initPython function and even the FFMpeg and Aria2c modules of the lib, but the YoutbeDL module itself can start to be reviewed.

JunkFood02 commented on library/src/main/java/com/yausername/youtubedl_android/util/plugins/PluginsUtil.kt:
Since we've learnt that the executables are required to be bundled in the apk, I think calling these zip files "dependencies" or something else should be more appropriate? "Plugins" implies that these files are optional for using this library

BobbyESP replied:
Renamed `Plugins` to `Dependencies`

JunkFood02 commented on library/src/main/java/com/yausername/youtubedl_android/YoutubeDL.kt:
```diff
@@ -16,84 +46,184 @@ import kotlin.collections.set
+     * @param callback a callback that will be called with the plugin and the progress of the download
+     */
+    @Throws(MissingPlugin::class)
+    fun assertPlugins(appContext: Context, downloadedPlugins: DownloadedPlugins, callback: pluginDownloadCallback?) {
```
A function called `assert` should do nothing other than assertions i think? maybe make it returns a boolean or throws an exception, then leave the rest to the caller

BobbyESP replied:
I've renamed the function. You tell me if it's okay

JunkFood02 commented on library/src/main/java/com/yausername/youtubedl_android/util/files/FilesUtil.kt:
```diff
@@ -0,0 +1,17 @@
+     * Asserts that a file exists and creates it if it doesn't
+     * @param file the file to assert or create in case it doesn't exist
+     */
+    fun assertAndCreate(file: File) {
```
creating a directory?

BobbyESP replied:
Renamed to `createDirectoryIfNotExists`

JunkFood02 commented on library/src/main/java/com/yausername/youtubedl_android/YoutubeDL.kt:
```diff
@@ -16,84 +46,184 @@ import kotlin.collections.set
+     * Initializes the library. This method should be called before any other method.
+     * @param appContext the application context
+     */
+    fun init(appContext: Context, callback: pluginDownloadCallback? = null) {
```
I prefer not to modify the previous APIs, especially this one, which might bring too many breaking changes.

JunkFood02 replied:
> We should maintain the original method of loading dependencies from the APK or external files, since network connections are not always reliable or simply for compatibility.

We can simply implement extract their common parts, and then separate the inconsistent logics/files into two methods or even two libraries

BobbyESP replied:
I have moved it to another method. In case that you want to move it to another library tell me; we have to modify both the FFMpeg module and Aria2c.
I was thinking on merging them into just the `library` module, so we can share the code. In other case we would have to make the thing you said before, making a separated library, but that would be very much job.

JunkFood02 replied:
Imagine the case that we need to build two variations of the same application, one using bundled-in dependencies and the other download them via internet connections. (Yes we do want to achieve this as I've mentioned before)

Spliting libraries solution: making the main `library` code able to handle these two cases, then package the native dependencies into standalone libraries. So we'll able to load the dependencies by simply include them in the gradle build scripts, or download them from specific URLs when the dependency libraries absent

BobbyESP replied:
Okay, so if I understood well the idea is to make the library be able to build in two different ways; one with the libraries already bundled, and another one without them, no?

JunkFood02 commented on library/src/main/java/com/yausername/youtubedl_android/YoutubeDL.kt:
```diff
@@ -254,20 +443,17 @@ object YoutubeDL {
         }
     }
 
+    internal val scope = CoroutineScope(SupervisorJob())
```
Be cautious to use coroutine in this library, or I'll just suggest avoiding coroutines. Because we don't know, and I believe we don't want to learn how coroutines behave in a Java world.

BobbyESP replied:
We have to use coroutines for Ktor; I tested the library download dependencies/plugins download in the Java example app and works perfect, no issues found. So I don't know if other solution can exist

BobbyESP replied:
I've deleted this value and added runBlocking in IO threads so the actual app thread isn't blocked, If you have any other alternative you tell me; by the moment gonna mark this review as resolved.

JunkFood02 commented on library/src/main/java/com/yausername/youtubedl_android/Constants.kt:
```diff
@@ -0,0 +1,33 @@
+        }
+    }
+
+    object Directories {
```
`DirectoryName.PYTHON` should make more sense (same for other constants)

BobbyESP replied:
Done!

BobbyESP committed:
refactor: Moved from `Plugin` to `Dependency`
refactor: Separated the dependencies ensure from the init function

BobbyESP committed:
feat: Added callback to app (as System.out)
fix: Dependency download with `installDependencies` updated every single iteration the progress

BobbyESP committed:
feat: Added callback to app (as System.out)
fix: Dependency download with `installDependencies` updated every single iteration the progress

BobbyESP commented:
Sorry for this last commit duplicated message [d4a2554](https://github.com/yausername/youtubedl-android/pull/256/commits/d4a25540a99b02812bb9543ebdc9720b930a85b6).

The changes were: 
- Delete the `scope` val from the library

BobbyESP committed:
Deleted unused test button on the MainActivity page

BobbyESP commented:
I hadn't a lot of time to continue on this, but will do. It seems to be more complex than what I thought, because also FFmpeg and Aria2c has to be downloadable, and by how the library is built we need to have some kind of shared code, meaning that I have to create another package with that shared code.

BobbyESP committed:
Process every line from the input stream not just the download-related

BobbyESP committed:
⚠️READ: Moved to the "common" package all the business logic of the dependencies downloader.
Now it's mandatory to implement the common package. A solution for this must be found

BobbyESP committed:
Merge remote-tracking branch 'origin/downloadable_packages' into downloadable_packages

BobbyESP committed:
refactor/WIP: Moving to Kotlin DSL and versions catalog

BobbyESP committed:
refactor/Gradle: Almost fully moved to Kotlin DSL and Versions Catalog

BobbyESP committed:
Update jitpack.yml to use Java 17

BobbyESP committed:
Update jitpack.yml - Deleted unnecessary before install commands

BobbyESP commented:
Created a [build](https://jitpack.io/#BobbyESP/youtubedl-android/623dcae771) so you can try how is this going. Far from a final release!!!

You can see an example [here](https://github.com/BobbyESP/youtubedl-android/blob/e9fdde29927653409c6188838d2fec673101e3f8/app/src/main/java/com/yausername/youtubedl_android_example/App.java#L59-L68)

BobbyESP committed:
Moved dependencies ensure to DependenciesUtil

BobbyESP committed:
Merge remote-tracking branch 'origin/downloadable_packages' into downloadable_packages

BobbyESP commented:
The `initPython ` function will have to be modified (as well as the other such as the FFmpeg and Aria2c) since the unzip process is already included in the download process. We should probably have to create different functions depending on the flavor

BobbyESP committed:
refactor: Created YoutubeDLCore class
⚠️(WIP): Created product flavors for `:library` module
- nonBundled version not finished and Android doesn't take it as a product variant

BobbyESP committed:
Now both versions working

BobbyESP commented:
@JunkFood02 Can we talk about how we should handle the libraries downloading? I mean. Should the download function be inside the `init` one or something like the already implemented in the app demo?

BobbyESP commented:
Seems to be an error at build time because of the app, here is the [log](https://jitpack.io/com/github/BobbyESP/youtubedl-android/f2a9355d8c/build.log). 

I think that the targetConfiguration isn't correct since I would say that the target should be "bundled" or "nonbundled" but when using those cases Gradle isn't able to find the library: 
```gradle
implementation project(":library") targetConfiguration("bundledDebugRuntimeElements")
```

BobbyESP committed:
- Fully moved to Kotlin DSL
- Added flavor to the test app
- Preparing the other dependencies for flavor-ization

BobbyESP committed:
(refactor): Moved `ensureDependencies` function to `YoutubeDLCore` as an abstract function

BobbyESP committed:
fix: Fixed `common` module having the same artifact ID as the `library` one

BobbyESP committed:
Moved FFmpeg to product flavors

BobbyESP committed:
Added binary files to VCS

BobbyESP committed:
Moved Aria2c to product flavor

JunkFood02 commented:
This PR is quite large (4000+ LoC) and hard to review. Could we please split it to smaller and more focused PRs, each addressing a specific topic?

BobbyESP commented:
The changes that are necessary for this functionality require all the work put into this PR since it means almost a rewrite of the library. I will still try to adjust the code to reduce the number of lines and changes, but please keep in mind that this proposed separation requires the changes made.

BobbyESP commented:
I will start by removing the Jetpack Compose app

JunkFood02 commented:
> The changes that are necessary for this functionality require all the work put into this PR

We can just make a pull request depending on another, which means a pull request could only be merge after all it's dependencies got merged into the target branch. This allows us to gradually applying the changes and review them in smaller chunks, making the process more manageable.

BobbyESP commented:
> > The changes that are necessary for this functionality require all the work put into this PR
> 
> We can just make a pull request depending on another, which means a pull request could only be merge after all it's dependencies got merged into the target branch. This allows us to gradually applying the changes and review them in smaller chunks, making the process more manageable.

That's interesting and cool. Can we create those chunks in some way or I have to create different branches with manual changes?

JunkFood02 commented:
> I have to create different branches with manual changes

I'm afraid you'll have to do so by hand since github doesn't support such cool pull requests while gitlab is offering [this as a paid feature](https://docs.gitlab.com/17.0/ee/user/project/merge_requests/dependencies.html)

You can start with building upon an existing branch. Submit a pull request when you're ready for a new, separate feature and create a new branch based of the current branch for further changes. Just make sure each branch(PR) focus on a single modification or topic for clarity.

Also, when developing on parallel branches simultaneously, avoid modifying the same files. You know what would happen 😈

JunkFood02 commented:
For example: you can start with the [`update_library`](https://github.com/yausername/youtubedl-android/tree/update_library) branch

1. migrating to Gradle Kotlin Script
2. update dependencies
3. other independent changes that could be working on at the same time
    - a new compose app for testing
    - replace fastxml with kotlinx.serialization
    - ...
4. decouple executables and dynamic-linked libraries (this PR)

BobbyESP commented:
Hahah okay, thanks for the help! I will do this as it should be done, as you told me. I'll keep this PR open for a while, but when I think the other smaller PRs cover everything done in this one, I'll close it.

deniscerri commented:
@BobbyESP The problem with chaquopy is that the packages are managed by them and alot of them are not available. Like curl cffi. Termux is our best bet i guess
Especially when we have to rely on other executables like ffmpeg.

Id love if we could gather up on discord or telegram, talk this through and try to build a system where packages are built through github actions. We code the shell scripts in the repo and the workflows execute them and then the app downloads them.

I introduced the idea to xibr, he mentioned its difficult but i think we can figure it out.

BobbyESP commented:
> @BobbyESP The problem with chaquopy is that the packages are managed by them and alot of them are not available. Like curl cffi. Termux is our best bet i guess Especially when we have to rely on other executables like ffmpeg.
> 
> Id love if we could gather up on discord or telegram, talk this through and try to build a system where packages are built through github actions. We code the shell scripts in the repo and the workflows execute them and then the app downloads them.
> 
> I introduced the idea to xibr, he mentioned its difficult but i think we can figure it out.

Hi mate!

Yeah, you're absolutely right. I personally prefer to discuss things over Telegram first, if that works for you. Thanks a lot for reaching out! I'd be happy to help, even though I’m no longer working on this, but it definitely sounds like an exciting challenge.

deniscerri commented:
we could make a gc in telegram, invite me
